File: jetbrains/src/main/kotlin/dev/pacmon/jetbrains/core/NugetManifest.kt
package dev.pacmon.jetbrains.core

object NugetManifestAdapter : ManifestAdapter {
    override val kind = ManifestKind.NUGET
    override val fileNames = listOf("Directory.Packages.props")
    override val discoveryGlobs = listOf("**/*.csproj", "**/*.fsproj", "**/*.vbproj", "**/Directory.Packages.props")
    override val notesRelativePath = ".pacmon/nuget/DEPENDENCY-NOTES.md"

    private val projectExtensions = listOf(".csproj", ".fsproj", ".vbproj")
    private val packageId = Regex("^[A-Za-z0-9][A-Za-z0-9._-]*$")

    override fun matchesPath(path: String): Boolean {
        val name = basename(path)
        val lower = name.lowercase()
        return name == "Directory.Packages.props" || projectExtensions.any(lower::endsWith)
    }

    override fun normalizeNoteKey(raw: String): String = ManifestRegistry.stripName(raw).lowercase()

    override fun extractDependencies(text: String): List<DependencyEntry> =
        extractDependencies(text, "project.csproj")

    override fun extractDependencies(text: String, path: String): List<DependencyEntry> {
        val nodes = descendants(parseXml(text))
        if (basename(path) == "Directory.Packages.props") {
            return nodes.flatMap { node ->
                when (node.name) {
                    "PackageVersion" -> {
                        val attribute = if (xmlAttribute(node, "Include") != null) "Include" else "Update"
                        entry(node, attribute, "centralVersion")?.let(::listOf).orEmpty()
                    }
                    "GlobalPackageReference" -> entry(node, "Include", "globalPackageReference")
                        ?.let(::listOf).orEmpty()
                    else -> emptyList()
                }
            }
        }
        return nodes.flatMap { node ->
            if (node.name != "PackageReference" || xmlAttribute(node, "Include") == null) emptyList()
            else entry(node, "Include", "packageReference")?.let(::listOf).orEmpty()
        }
    }

    private fun entry(node: XmlNode, attributeName: String, scope: String): DependencyEntry? {
        if (xmlAttribute(node, "Remove") != null) return null
        val attribute = xmlAttribute(node, attributeName) ?: return null
        if (!packageId.matches(attribute.value)) return null
        return dependencyEntry(
            attribute.value,
            scope,
            attribute.range,
            listOf(attribute.range),
            attribute.value,
            SourceRange(node.openStart, 1),
        )
    }

    private fun descendants(roots: List<XmlNode>): List<XmlNode> = buildList {
        fun visit(node: XmlNode) {
            add(node)
            node.children.forEach(::visit)
        }
        roots.forEach(::visit)
    }

    private fun basename(path: String): String = path.replace('\\', '/').substringAfterLast('/')
}

File: src/core/nuget-manifest.ts
import { dependencyEntry } from './dependency';
import type { DependencyEntry } from './model';
import { parseXml, xmlAttribute, type XmlNode } from './xml';

const PROJECT_EXTENSIONS = ['.csproj', '.fsproj', '.vbproj'] as const;
const NUGET_PACKAGE_ID = /^[A-Za-z0-9][A-Za-z0-9._-]*$/;

function basename(path: string): string {
  return path.replace(/\\/g, '/').split('/').at(-1) ?? '';
}

export function isNugetManifestPath(path: string): boolean {
  const name = basename(path);
  const lower = name.toLowerCase();
  return name === 'Directory.Packages.props'
    || PROJECT_EXTENSIONS.some((extension) => lower.endsWith(extension));
}

function descendants(nodes: readonly XmlNode[]): XmlNode[] {
  const result: XmlNode[] = [];
  const visit = (node: XmlNode): void => {
    result.push(node);
    node.children.forEach(visit);
  };
  nodes.forEach(visit);
  return result;
}

function entry(node: XmlNode, attributeName: 'Include' | 'Update', scope: string): DependencyEntry | undefined {
  if (xmlAttribute(node, 'Remove')) return undefined;
  const attribute = xmlAttribute(node, attributeName);
  if (!attribute || !NUGET_PACKAGE_ID.test(attribute.value)) return undefined;
  const iconRange = { offset: node.openStart, length: 1 };
  return dependencyEntry(
    attribute.value,
    scope,
    attribute.range,
    [attribute.range],
    attribute.value,
    iconRange,
  );
}

/** Reads literal NuGet declarations without evaluating MSBuild or following imports. */
export function extractNugetDependencies(text: string, path = 'project.csproj'): DependencyEntry[] {
  const nodes = descendants(parseXml(text));
  if (basename(path) === 'Directory.Packages.props') {
    return nodes.flatMap((node) => {
      if (node.name === 'PackageVersion') {
        const attributeName = xmlAttribute(node, 'Include') ? 'Include' : 'Update';
        const dependency = entry(node, attributeName, 'centralVersion');
        return dependency ? [dependency] : [];
      }
      if (node.name === 'GlobalPackageReference') {
        const dependency = entry(node, 'Include', 'globalPackageReference');
        return dependency ? [dependency] : [];
      }
      return [];
    });
  }

  return nodes.flatMap((node) => {
    if (node.name !== 'PackageReference' || !xmlAttribute(node, 'Include')) return [];
    const dependency = entry(node, 'Include', 'packageReference');
    return dependency ? [dependency] : [];
  });
}

/** Coverage uses one row per case-insensitive NuGet package ID; callers pass
 * central manifests first so their declaration represents duplicates. */
export function uniqueNugetDependencies(dependencies: readonly DependencyEntry[]): DependencyEntry[] {
  const unique = new Map<string, DependencyEntry>();
  for (const dependency of dependencies) {
    const key = dependency.noteKey.trim().toLowerCase();
    if (!unique.has(key)) unique.set(key, dependency);
  }
  return [...unique.values()];
}
